service/history/ndc/history_replicator.go

```diff
@@ -227,6 +229,193 @@ func (r *HistoryReplicatorImpl) ApplyEvents(
 	return r.doApplyEvents(ctx, task)
 }
 
+func (r *HistoryReplicatorImpl) BackfillHistoryEvents(
+	ctx context.Context,
+	request *shard.BackfillHistoryEventsRequest,
+) error {
+	task, err := newReplicationTaskFromBatch(
+		r.clusterMetadata,
+		r.logger,
+		request.WorkflowKey,
+		request.BaseExecutionInfo,
+		request.VersionHistoryItems,
+		request.Events,
+		request.NewEvents,
+		request.NewRunID,
+		request.VersionedHistory,
+	)
+	if err != nil {
+		return err
+	}
+
+	return r.doApplyBackfillEvents(ctx, task, r.applyBackfillEvents)
+}
+
+func (r *HistoryReplicatorImpl) doApplyBackfillEvents(
+	ctx context.Context,
+	task replicationTask,
+	action func(context.Context, workflow.MutableState, workflow.Context, wcache.ReleaseCacheFunc, replicationTask) error,
+) (retError error) {
+	wfContext, releaseFn, err := r.workflowCache.GetOrCreateWorkflowExecution(
+		ctx,
+		r.shardContext,
+		task.getNamespaceID(),
+		task.getExecution(),
+		locks.PriorityHigh,
+	)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if rec := recover(); rec != nil {
+			releaseFn(errPanic)
+			panic(rec)
+		}
+		releaseFn(retError)
+	}()
+
+	mutableState, err := wfContext.LoadMutableState(ctx, r.shardContext)
+	switch err.(type) {
+	case nil:
+		return action(ctx, mutableState, wfContext, releaseFn, task)
+	case *serviceerror.NotFound:
+		return serviceerrors.NewSyncState(
+			mutableStateMissingMessage,
+			task.getNamespaceID().String(),
+			task.getWorkflowID(),
+			task.getRunID(),
+			task.getVersionedTransition(),
+		)
+	default:
+		return err
+	}
+}
+
+func (r *HistoryReplicatorImpl) applyBackfillEvents(
+	ctx context.Context,
+	mutableState workflow.MutableState,
+	wfContext workflow.Context,
+	releaseFn wcache.ReleaseCacheFunc,
+	task replicationTask,
+) (retError error) {
+
+	versionedTransition := task.getVersionedTransition()
+	if versionedTransition == nil {
+		return errors.New("versioned transition is required for backfill task")
+	}
+
+	if task.getFirstEvent().GetEventType() == enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_STARTED {
+		return errors.New("workflow execution started event is not expected in backfill task")
+	}
+
+	transitionHistory := mutableState.GetExecutionInfo().GetTransitionHistory()
+	err := workflow.TransitionHistoryStalenessCheck(transitionHistory, versionedTransition)
```

Comment:
we want to backfill a versioned transition on a different branch so we expect the staleness check to return an error.

I think we need to compare the versioned transition in the replication task to the one in mutable state with `CompareVersionedTransition` 
- if the one in replication task is larger, return syncState error
- otherwise, use `TransitionHistoryStalenessCheck` to see if the versioned transition the task is trying to backfill is on the current transition branch. If so (no error returned), we don't need to do anything, otherwise (no error returned), start the backfill process.

Reply:
updated